src/charm.py

```diff
@@ -0,0 +1,192 @@
+        evaluation = self.kube_control.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        if not self.kube_control.get_auth_credentials(self.unit.name):
+            self.unit.status = WaitingStatus("Waiting for kube-control: unit credentials.")
+            return False
+        self.unit.status = MaintenanceStatus("Kubernetes authentication completed.")
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/root/.kube/config", "root", self.unit.name
+        )
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/home/ubuntu/.kube/config", "ubuntu", self.unit.name
+        )
+        return True
+
+    def _check_certificates(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self.CA_CERT_PATH.write_text(self.certificates.ca)
+        return True
+
+    def _check_config(self, event: EventBase) -> None:  # TODO: Not the best name
+        if not self._check_certificates(event):
+            return
+
+        if not self._check_kube_control(event):
+            return
+
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+        self.unit.status = ops.ActiveStatus("Ready to test.")
+
+    def _upgrade_charm(self, event: EventBase) -> None:
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+    def _install_snaps(self, channel: Optional[str]) -> None:
+        self.unit.status = ops.MaintenanceStatus("Installing core snap.")
+        self._ensure_snap("core")
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubectl snap.")
+        self._ensure_snap("kubectl", channel=channel)
+
+        self.unit.status = ops.MaintenanceStatus("Installing kubernetes-test snap.")
+        self._ensure_snap("kubernetes-test", channel=channel, classic=True)
+
+    def _ensure_snap(
+        self,
+        name: str,
+        state: SnapState = SnapState.Latest,
+        channel: Optional[str] = "",
+        classic: Optional[bool] = False,
+    ) -> None:
+        if not isinstance(name, str) or name == "":
+            raise ValueError("A name is required to ensure a snap.")
+
+        snap = self.snap_cache[name]
+        if not snap.present:
+            snap.ensure(state=state, classic=classic, channel=channel)
+
+    def _on_start(self, event: EventBase) -> None:
+        self.unit.status = ops.ActiveStatus()
```

Comment:
Nah, let's not state the charm is active just b/c the first hook fired.  That sends a bad message